sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +303,166 @@ def get_lpmode_support(self):
+                time.sleep(0.5)
+                val = self.xcvr_eeprom.read(consts.MODULE_LEVEL_CTRL_RESET_FIELD)
+                # Device is in REST, the I2C read could fail, and None could be returned
+                if (val is None) or val:
+                    ret = False
+                    continue
+                # Device is out of REST
+                ret = True
+                break
+        return ret
+
+    def get_error_description(self):
+        val = self.xcvr_eeprom.read(consts.DATAPATH_STATE_FIELD)
+        if val not in [0, CMIS_DATAPATH_STATE_ACTIVATED]:
+            return SFP_ERROR_DESCRIPTION_CMIS_DATAPATH_ERROR
+
+        val = self.xcvr_eeprom.read(consts.CONFIG_ERROR_FIELD)
+        if val not in [0, CMIS_CONFIG_STATE_ACCEPTED]:
+            return SFP_ERROR_DESCRIPTION_CMIS_CONFIG_ERROR
+
+        state = self.xcvr_eeprom.read(consts.MODULE_STATE_FIELD)
+        if state not in [CMIS_MODULE_STATE_NONE, CMIS_MODULE_STATE_READY]:
+            return SFP_ERROR_DESCRIPTION_CMIS_NOT_READY
+
+        return None
+
+    def get_host_speed(self, ifname):
+        # see host_electrical_interface of sff8024.py
+        speed = 0
+        if '400G' in ifname:
+            speed = 400000
+        elif '200G' in ifname:
+            speed = 200000
+        elif '100G' in ifname or 'CAUI-4' in ifname:
+            speed = 100000
+        elif '50G' in ifname or 'LAUI-2' in ifname:
+            speed = 50000
+        elif '40G' in ifname or 'XLAUI' in ifname or 'XLPPI' in ifname:
+            speed = 40000
+        elif '25G' in ifname:
+            speed = 25000
+        elif '10G' in ifname or 'SFI' in ifname or 'XFI' in ifname:
+            speed = 10000
+        elif '1000BASE' in ifname:
+            speed = 1000
+        return speed
+
+    def get_cmis_application_selected(self):
+        ap_code = 0
+        if not self.is_flat_memory():
+            field = "{}_{}".format(consts.STAGED_CTRL0_APSEL_FIELD, 1)
+            ap_code = self.xcvr_eeprom.read(field) >> 4
+
+        return (ap_code & 0xf)
+
+    def get_cmis_application_matched(self, host_speed, host_lanes):
+        if host_speed == 0 or host_lanes == 0:
+            return 0
+
+        ap_code = 0
+        appl_dict = ast.literal_eval(self.xcvr_eeprom.read(consts.APPLS_ADVT_FIELD))
+        for c in appl_dict.keys():
+            d = appl_dict[c]
+            if d.get('host_lane_count') != host_lanes:
+                continue
+            if self.get_host_speed(d.get('host_electrical_interface_id')) != host_speed:
+                continue
+            ap_code = c
+
+        return (ap_code & 0xf)
+
+    def set_cmis_application(self, host_speed, host_lanes):
+        if self.is_flat_memory():
+            return False
+
+        # Only applicable to CMIS v4.0 onwards
+        if self.xcvr_eeprom.read(consts.CMIS_REVISION_FIELD) < 0x40:
+            return False
+
+        # Check for application selection updates
```

Comment:
Sorry, I'm not following you. Did you mean a re-initialization is necessary even when the module/lane is in default application mode (i.e. Both ConfigError and DataPathInitState are 0x00)? If that's the case, we could update the checker NOT to skip this operation.
But if it does not seem to be necessary to do re-initialization if the module/lane is already correctly configured and passing traffic while no port mode updates.

Reply:
This logic will only ever be triggered on port speed or breakout change (ie: traffic is going to be disrupted anyway). Performing re-initialization ensures that the module adapation is done after the ASIC configuration has been completed.

Reply:
1. The the latest commit, only default application (ApSel = 1) is supported, we'll deal with port breakout support once the synchronization mechanism between syncd and xcvrd is in place and ready.
2. In the latest code change, in the case that CMIS is initialized by the hardware init sequence, the ConfigError will be 0 (No status), and it will trigger the CMIS software init sequence. In other words, the CMIS will always be re-initialized with the software init sequence.